### PR TITLE
ACR Ghosting (and Uniformity)

### PR DESCRIPTION
**Pull request by YassineAzma**
- Adds an ACR ghosting script and associated unit tests that calculates the percent signal ghosting of slice 7 of the ACR phantom. This is done within the new HazenTask framework. The methodology is as follows:
  - Calculates the centroid of the object following morphological operations and a centre-of-mass calculation.
    - This step also produces a binary mask.
  - Define a large ROI, of 200cm<sup>2</sup> area and offset from the top of the phantom. 
  - Elliptical ROIs of 10cm<sup>2</sup> area are then drawn between the edges of the FoV and the phantom along the cardinal directions. These ROIs are stretched along the long-axis of the ellipses depending on how close they are to the FoV edge and phantom edge. The 10cm<sup>2</sup> area of each ellipse is maintained.
  - The mean values from each ROI are used to calculate the percent-signal ghosting. Which is defined as: $$PSG=100* {|(north + south) - (west + east)|\over(2*large ROI)} $$
  - These results are visualised in the report, as shown below:
ACR_T1_TRA_HEAD_7_7

**Additional edits**
- Refactored ACR uniformity script and associated unit tests to work within the HazenTask class.
- Associated modifications to __init__.py to (re)add the necessary hazen CLI functions for the ACR uniformity and ghosting scripts.

N.B: There is no existing issue for the ACR ghosting test.

### CODE REVIEW

File: hazenlib/__init__.py [hunk 1]
```diff
@@ -383,6 +384,10 @@ def main():
         task = importlib.import_module(f"hazenlib.{arguments['<task>']}")
         dicom_objects = [pydicom.read_file(x, force=True) for x in files if is_dicom_file(x)]
         result = task.main(dicom_objects, report_path=True)
+    elif arguments['<task>'] == 'acr_ghosting':
+        task = importlib.import_module(f"hazenlib.{arguments['<task>']}")
+        dicom_objects = [pydicom.read_file(x, force=True) for x in files if is_dicom_file(x)]
+        result = task.main(dicom_objects, report_path=True)
```
Comment (Lucrezia-Cester) [inline hazenlib/__init__.py]: I would remove these two statements

File: hazenlib/tasks/task_acr_ghosting.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+        sum_x = np.sum(coords[1])
+        sum_y = np.sum(coords[0])
+        cxy = sum_x / coords[0].shape, sum_y / coords[1].shape
+
+        cxy = [cxy[0].astype(int), cxy[1].astype(int)]
+        return bhull, cxy
+
+    def get_signal_ghosting(self, dcm):
+        img = dcm.pixel_array  # extract image slice of interest
```
Comment (Lucrezia-Cester) [inline hazenlib/tasks/task_acr_ghosting.py]: What does the comment refer to? Is this not just taking the image values?

File: hazenlib/tasks/task_acr_ghosting.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    def run(self) -> dict:
+        results = {}
+        z = []
+        for dcm in self.data:
+            z.append(dcm.ImagePositionPatient[2])
+
+        idx_sort = np.argsort(z)
+
+        for dcm in self.data:
+            if dcm.ImagePositionPatient[2] == z[idx_sort[6]]:
```
Comment (Lucrezia-Cester) [inline hazenlib/tasks/task_acr_ghosting.py:41]: This line throws an error why I tested this with the ghosting test data on Hazen --->

index 6 is out of bounds for axis 0 with size 2

Comment (YassineAzma) [reply hazenlib/tasks/task_acr_ghosting.py:41]: You need to use the test data in the ACR folder. My scripts are designed to be ran on a folder that contains all 11 slices produced by sequences used for ACR QA. The ghosting test only looks at the seventh slice.

File: hazenlib/tasks/task_acr_ghosting.py (hunk 2 above)
Comment (YassineAzma) [reply hazenlib/tasks/task_acr_ghosting.py]: That's an old comment from my MATLAB script where I would work directly from a 3D array. I'll remove that - what that line does is apparent anyway.

File: hazenlib/__init__.py (hunk 1 above)
Comment (YassineAzma) [reply hazenlib/__init__.py]: Sure.

@Lucrezia-Cester 
Question: I wanted to keep the python scripts as acr_ghosting and acr_uniformity (without task_ at the beginning), but even after deleting my old scripts they were still being called in the command line. Anything I delete seems to be rebuilt in the virtual environment.

File: hazenlib/tasks/task_acr_ghosting.py (hunk 3 above)
Comment (Lucrezia-Cester) [reply hazenlib/tasks/task_acr_ghosting.py:41]: @YassineRMH, checked with the ACR data and it works, thanks!

**Comment by YassineAzma** [comment]
@Lucrezia-Cester Do you know what this acr_snr CLI function that keeps failing is? I didn't include that, so I have no idea. 

Edit: I've went through Laurence's pull request for the HazenTask refactoring and it seems to be what he renamed the acr_uniformity task to. I'll delete it.

**Comment by Lucrezia-Cester** [comment]
Hi @YassineRMH thank you so much for fixing that, I'll check the uniformity script and then approve this PR soon

**Comment by Lucrezia-Cester** [comment]
Hi @YassineRMH, by the way, feel free to add issues, and in general it would be best to separate different tasks in different PRs, in this case, ghosting from uniformity

**Comment by YassineAzma** [comment]
> Hi @YassineRMH, by the way, feel free to add issues, and in general it would be best to separate different tasks in different PRs, in this case, ghosting from uniformity

Will do.

**Review by Lucrezia-Cester — DISMISSED** [review]

**Comment by Lucrezia-Cester** [comment]
Hi @YassineRMH, before I merge, would it be possible for you to add the docstirng for the unifromity script the same way you did for the ghosting? I am keen to keep track of sources.

**Comment by YassineAzma** [comment]
@Lucrezia-Cester Potentially stupid question, but what do you mean by docstring?

Edit: Done.

**Comment by Lucrezia-Cester** [comment]
Hi @YassineRMH, is there a link to the ACR guidance?

**Comment by YassineAzma** [comment]
@Lucrezia-Cester Do you want me to include a link to the guidance in my docstrings?

If not, it's here:

https://www.acraccreditation.org/-/media/acraccreditation/documents/mri/largephantomguidance.pdf

**Comment by Lucrezia-Cester** [comment]
Hi @YassineRMH, i just added it, I'll merge now

**Comment by YassineAzma** [comment]
@Lucrezia-Cester One last thing, I was trying to edit this before you merged:

On line 6 in the task_acr_ghosting.py script, could you please change uniformity to percent-signal ghosting. I can also make a separate pull request for this, but it seems very overkill.

**Comment by Lucrezia-Cester** [comment]
Hi @YassineRMH, apologies for not waiting for this edit.  Will change it myself.

**Comment by YassineAzma** [comment]
@Lucrezia-Cester Thanks for all your help. 👍

**Pull request merged**